Fix windows line endings#791 - Windows delimiter is accepted as standard

tobwiens opened the pull request:
This is a fix for issue: #791 

If the standard delimiter is \n, a \r\n is accepted as well. Therefore a
marker was introduced to indicate that specific case. The delimiter
length is variably set to the windows delimiter size then. To prevent
failing in reading input with windows line endings.
The DelimitedInputFormat.open() method did not check if it has parsers
registered. Added throwing an exception when no parser is registered
when calling open. CsvInputFormat does now accept windows line endings
with standards settings. The |\r\n" will be replaced by "\n_"( _ is a
space)... To prevent from errors the returned length is set to the
space. So that this byte is known as parsed and will not get interpreted
anymore.

tobwiens committed:
Exception added to open() and fixed \r\n

The DelimitedInputFormat.open() method did not check if it has parsers
registered. Added throwing an exception when no parser is registered
when calling open. CsvInputFormat does now accept windows line endings
with standards settings. The |\r\n" will be replaced by "\n_"( _ is a
space)... To prevent from errors the returned length is set to the
space. So that this byte is known as parsed and will not get interpreted
anymore.

tobwiens committed:
Example accesses array out of bounds

args.length >= 4 checks if a fourth element is in the array but then it
accesses [4], the fifth element.

tobwiens committed:
Windows delimiter is accepted as standard

If the standard delimiter is \n, a \r\n is accepted as well. Therefore a
marker was introduced to indicate that specific case. The delimiter
length is variably set to the windows delimiter size then. To prevent
failing in reading input with windows line endings.

tobwiens committed:
Checkstyle compliance

if without { } was used.

tobwiens commented:
Is not stable yet.

tobwiens committed:
Revert "Windows delimiter is accepted as standard"

This reverts commit 040b06ff316c423ddf3cf5df4e0f1164a50def07.

tobwiens committed:
Bring windows line ending check into CsvInputFor.

The windows line ending support is brought into the CsvInputFormat.java
from the DelimitedInputFormat.... Because that created failed test cases
from the terra sort benchmark

tobwiens committed:
Remove not needed final static variables

tobwiens committed:
Detect line ending before parsing

The numBytes is reduced when a windows file ending is found.

tobwiens committed:
Better comments

StephanEwen commented on stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CsvInputFormat.java:
```diff
@@ -93,26 +98,50 @@ public void open(FileInputSplit split) throws IOException {
+		
+		//left to right evaluation makes access [0] okay
+		//this marker is used to fasten up readRecord, so that it doesn't have to check each call if the line ending is set to default
+		if(this.getDelimiter().length == 1 && this.getDelimiter()[0] == "\n".getBytes(Charsets.UTF_8)[0] ) {
```
I think you can simplify this to `getDelimiter()[0] == '\n'`

I think that line feed is the same in all charsets that matter. We do exclude mainframe and C64 charsets that way, but I am okay with that ;-)

StephanEwen commented on stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CsvInputFormat.java:
```diff
@@ -93,26 +98,50 @@ public void open(FileInputSplit split) throws IOException {
+		 * Fix to support windows line endings in CSVInputFiles with standard delimiter setup = \n
+		 */
+		//Find windows end line, so find chariage return before the newline 
+		if(this.lineDelimiterIsLinebreak == true && bytes[offset + numBytes -1] == "\r".getBytes(Charsets.UTF_8)[0]) {
```
`"\r".getBytes(Charsets.UTF_8)[0]` is a really expensive call. It goes through the charset codecs.
Just do a `bytes[offset + numBytes -1] == '\r'`.

tobwiens committed:
Checks done faster

exchanged by a less expansive check.

tobwiens committed:
Same changes introduced for record\io

The class in record\io gets the same changes.

tobwiens committed:
Introducing test cases record\io

Adding test case for CsvInputFormat, which check the correct windows
line handling.

tobwiens committed:
Removed unused import

tobwiens committed:
Adapting test cases to record\io CsvInputFormat

tobwiens commented:
Thank you Stefan, I addressed the issues!

I left the length == 1 check, to not interfere with arbitrary delimiters which might contain a \n at position 0.

StephanEwen commented on stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CsvInputFormat.java:
```diff
@@ -93,26 +97,50 @@ public void open(FileInputSplit split) throws IOException {
+		//Find windows end line, so find chariage return before the newline 
+		if(this.lineDelimiterIsLinebreak == true && bytes[offset + numBytes -1] == '\r' ) {
+			//reduce the number of bytes so that the Carriage return is not taken as data
+			numBytes--;
```
I think it would be good to add a check if that reduces teh line to length 0. Skip the line in that case.

StephanEwen commented:
Fixed up and merged in b0557d08865d31762fb2ddb913b4aa7a2f3ed3b3
